### PR TITLE
[JIT] Add property support to TorchScript classes

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42390 [JIT] Add property support for ScriptModules
* **#42389 [JIT] Add property support to TorchScript classes**

**Summary**
This commit adds support for properties to TorchScript classes,
specifically for getters and setters. They are implemented essentially
as pointers to the methods that the corresponding decorators decorate,
which are treated like regular class methods. Deleters for properties
are considered to be out of scope (and probably useless for TorchScript
anyway).

**Test Plan**
This commit adds a unit test for a class with a property that has both
getter and setter and one that has only a getter.

`python test/test_jit.py TestClassType.test_properties`

Differential Revision: [D22880232](https://our.internmc.facebook.com/intern/diff/D22880232)

### CODE REVIEW

File: torch/csrc/jit/frontend/sugared_value.cpp [hunk 1]
```diff
@@ -284,6 +290,16 @@ void SimpleValue::setAttr(
             << "Initialize the field at the top level first";
       }
     } else {
+      // Check and see if it's a setter attribute.
+      auto prop = classType->getProperty(field);
+      if (prop && prop->setter) {
+        auto& g = *m.graph();
```
Comment (eellison) [inline torch/csrc/jit/frontend/sugared_value.cpp:296]: Could you refactor this to use `MethodValue(value_, getter_name)->call(...)` ? If you look at the function there is additional logic there like checking for recursive calls, and setting the source range location that is not being done here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/sugared_value.h#L448

File: torch/csrc/jit/python/script_init.cpp [hunk 2]
```diff
@@ -1303,11 +1305,21 @@ void initJitScriptBindings(PyObject* module) {
                    "something else!";
           }
           methodDefs.emplace_back(Def(def));
-          rcbs.push_back(
+          methodRcbs.push_back(
               pythonResolver(rcb, classDef.name().name(), classType));
         }
+
+        // Compile getters and setters for properties as regular methods.
```
Comment (eellison) [inline torch/csrc/jit/python/script_init.cpp:1312]: nit: maybe rephrase this comment ? not sure i'm parsing it well

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 3]
```diff
@@ -3793,6 +3826,75 @@ std::unique_ptr<Function> CompilationUnit::define(
...
+
+    if (setter_fn) {
+      const auto& setter_name = setter_fn->name();
+      function_table[setter_name] = setter_fn.get();
+      functions.push_back(setter_fn.get());
+      register_function(std::move(setter_fn));
+    }
+  }
+
+  for (size_t i = 0; i < definitions.size(); i++) {
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3869]: this is a copy-pasta from the other define function, can you unify this and the existing one into one implementation ?

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 4]
```diff
@@ -3793,6 +3826,75 @@ std::unique_ptr<Function> CompilationUnit::define(
...
+            properties[i],
+            propResolvers[i],
+            self,
+            function_table,
+            shouldMangle);
+
+    auto& getter_fn = property_fns.first;
+    auto& setter_fn = property_fns.second;
+
+    const auto& getter_name = getter_fn->name();
```
Comment (eellison) [inline torch/csrc/jit/frontend/ir_emitter.cpp:3856]: This four-line block of code gets repeated three times in this function, can we factor it out ?

File: test/jit/test_class_type.py [hunk 5]
```diff
@@ -1027,3 +1027,63 @@ def foo():
...
+        Test that a scripted class can make use of the @property decorator.
+        """
+        @torch.jit.script
+        class Properties(object):
+            def __init__(self, a: int):
+                self.a = a
+
+            @property
+            def attr(self) -> int:
+                return self.a
```
Comment (eellison) [inline test/jit/test_class_type.py:1042]: for the getters/setters maybe add some subtraction/addition to stress test this a bit... not that it should rreally matter

**Review by eellison — COMMENTED** [review]
Cool, this looks great ! I think this is logically good to go, but there are a couple refactorings that would be good to do before landing. Those should be pretty mechanical/easy to do though.

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 6]
```diff
@@ -158,6 +158,8 @@ c10::optional<std::string> ScriptTypeParser::parseBaseTypeName(
       // Special case for torch.Tensor
       if (isTorch(select.value()) && name == "Tensor") {
         return "Tensor";
+      } else if (isTorch(select.value()) && name == "device") {
+        return "Device";
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/script_type_parser.cpp:162]: Internal fbcode Detectron2 tests break without it because they use `-> torch.device` as a return type annotation.

File: torch/csrc/jit/python/script_init.cpp (hunk 2 above)
Comment (SplitInfinity) [reply torch/csrc/jit/python/script_init.cpp:1312]: It's no longer accurate in this version, I'll rewrite it.

File: torch/csrc/jit/frontend/script_type_parser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/frontend/script_type_parser.cpp:162]: That should work bc of https://github.com/pytorch/pytorch/blob/master/torch/jit/annotations.py#L313. I think what's happening here is the resolver for properties is not being set properly. We should be closing over `torch` bc it's used in the function, but we're not. 

Sometimes with python wrapped functions you have to unpack the wrapped function to do resolution correctly, as in https://github.com/pytorch/pytorch/pull/40468. 

Could you add a test where a property getterr function calls into a free-function ? 
```
def foo(x):
    return x + 1
...

@property
def attr(self):
      return foo(self.a)
```

that should check whether the resolver is being properly set

[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/frontend/script_type_parser.cpp:162]: short repro:
```
class Boxes(object):
    @property
    def device(self) -> torch.device:
        return 1 + hi


closure = torch._jit_internal.get_closure(Boxes.device.fget)
# torch is closed over
print(closure)
```

File: aten/src/ATen/core/type.cpp [hunk 7]
```diff
@@ -1341,6 +1341,21 @@ std::shared_ptr<const CompilationUnit> ClassType::compilation_unit() const {
...
+    if (name == prop.name) {
+      return prop;
+    }
+  }
+
+  return c10::nullopt;
+}
+
+void ClassType::addProperty(const std::string& name, torch::jit::Function* getter, torch::jit::Function* setter) {
+  properties_.push_back({name, getter, setter});
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply aten/src/ATen/core/type.cpp:1360]: Done.

File: torch/csrc/jit/api/compilation_unit.h [hunk 8]
```diff
@@ -83,6 +84,16 @@ struct TORCH_API CompilationUnit {
     return true;
   }
 
+  // Version of define but with support for properties.
+  std::vector<Function*> define(
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/api/compilation_unit.h]: Done. Although there are a bunch of callsites now where we compile free functions but still have to add empty `{}` for properties/property resolvers. I thought of adding them to the end of the argument list so as to avoid modifying all those callsites but it didn't seem as clean.

File: torch/csrc/jit/api/compilation_unit.h [hunk 9]
```diff
@@ -262,6 +273,14 @@ struct TORCH_API CompilationUnit {
       const std::unordered_map<std::string, Function*>& function_table,
       bool shouldMangle = false) const;
 
+  std::pair<std::unique_ptr<Function>, std::unique_ptr<Function>> define(
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/api/compilation_unit.h]: Done.

File: torch/csrc/jit/frontend/script_type_parser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/script_type_parser.cpp:162]: I added a test for this but I'm not sure if it right...

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 10]
```diff
@@ -3746,6 +3746,39 @@ CompilationUnit::CompilationUnit(const std::string& source)
...
+        const std::unordered_map<std::string, Function*>& function_table,
+        bool shouldMangle) const {
+  TORCH_INTERNAL_ASSERT(self);
+  TORCH_INTERNAL_ASSERT(resolver);
+
+  std::unique_ptr<Function> getter_fn = define(
+      prefix, prop.getter(), resolver, self, function_table, shouldMangle);
+
+  std::unique_ptr<Function> setter_fn = prop.setter().present()
+      ? define(
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Done.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 11]
```diff
@@ -3746,6 +3746,39 @@ CompilationUnit::CompilationUnit(const std::string& source)
...
+std::pair<std::unique_ptr<Function>, std::unique_ptr<Function>> CompilationUnit::
+    define(
+        const c10::optional<c10::QualifiedName>& prefix,
+        const Property& prop,
+        const ResolverPtr& resolver,
+        const Self* self,
+        const std::unordered_map<std::string, Function*>& function_table,
+        bool shouldMangle) const {
+  TORCH_INTERNAL_ASSERT(self);
+  TORCH_INTERNAL_ASSERT(resolver);
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Removed.

File: torch/csrc/jit/frontend/ir_emitter.cpp [hunk 12]
```diff
@@ -3746,6 +3746,39 @@ CompilationUnit::CompilationUnit(const std::string& source)
...
+            prop.setter().get(),
+            resolver,
+            self,
+            function_table,
+            shouldMangle)
+      : nullptr;
+
+  auto* getter = self->getClassType()->findMethod(prop.getter().name().name());
+  auto* setter = prop.setter().present()
+      ? self->getClassType()->findMethod(prop.setter().get().name().name())
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Yeah, we can reuse them. I didn't realize that the return `findMethod` in these cases will return the same value as `getter_fn.get()` and `setter_fn.get()`.

File: torch/csrc/jit/frontend/sugared_value.cpp (hunk 1 above)
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/sugared_value.cpp]: Done.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 3 above)
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/ir_emitter.cpp:3894]: Done.

File: torch/csrc/jit/frontend/ir_emitter.cpp (hunk 4 above)
Comment (SplitInfinity) [reply torch/csrc/jit/frontend/ir_emitter.cpp]: Factored it out into a local lambda.

File: test/jit/test_class_type.py (hunk 5 above)
Comment (SplitInfinity) [reply test/jit/test_class_type.py]: Done.

**Comment by SplitInfinity** [comment]
Hey, I addressed all of the comments but I had to amend my commit because I didn't want `ghstack` to make another PR.

File: torch/csrc/jit/frontend/script_type_parser.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/csrc/jit/frontend/script_type_parser.cpp:162]: No, the test looks good. We shouldn't need to add the torch.device thing, it means something else has gone wrong... It would be nice to fix but if resolution in general is working, then I think we can land this. 

Could you remove this "torch.device" logic here, and then in the D2 test change the property to look like ? 
```
from torch import device

@property
def foo() -> device:
    ...
```
This should workaround the problem without added unnecessary stuff to our code base...

**Review by eellison — APPROVED** [review]
Sweet, looks great!! One comment about the torch.device thing.

Also, it would be maybe be nice to land this with the property support for TorchScript modules...

**Comment by SplitInfinity** [comment]
Okay, I reverted the `torch.device` change so that we can re-examine the failure. I held off on addressing comments on the properties for Modules PR until all of the major changes in this PR were acceptable in order to avoid spending too much time on rebasing. But now that this PR is approved, I will address comments on #42390 so that we can land them together or in quick succession.
